compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/irCodegenUtils.kt

```diff
@@ -418,23 +411,20 @@ internal fun getSignature(
    For other cases use getVisibilityAccessFlag(MemberDescriptor descriptor)
    Classes in byte code should be public or package private
 */
-fun IrClass.getVisibilityAccessFlagForClass(): Int {
-    /* Original had a check for SyntheticClassDescriptorForJava, never invoked in th IR backend. */
-    if (isOptionalAnnotationClass()) {
-        return AsmUtil.NO_FLAG_PACKAGE_PRIVATE
+fun IrClass.getVisibilityAccessFlagForClass(context: JvmBackendContext): Int =
+    when {
+        isOptionalAnnotationClass() -> AsmUtil.NO_FLAG_PACKAGE_PRIVATE
+        kind === ClassKind.ENUM_ENTRY -> AsmUtil.NO_FLAG_PACKAGE_PRIVATE
+        visibility === Visibilities.PRIVATE -> AsmUtil.NO_FLAG_PACKAGE_PRIVATE
+        isClassForCallableReference -> getVisibilityAccessFlagForAnonymous(context)
+        else -> Opcodes.ACC_PUBLIC
```

Comment:
I'd prefer to list all supported visibilities explicitly (`PUBLIC`, `PROTECTED`, `LOCAL`, `INTERNAL`), keeping the TODO about local visibility and reflection, and throw ISE otherwise if the visibility is unknown. This will allow this code to fail faster in case we use some new visibility (such as `JavaVisibilities.PACKAGE_VISIBILITY`) in some new lowering

Reply:
Done! And a good thing, too, since it turns out that there are cases where we have classes with package visibility...